Utils and libraries: Fix #13371 - move umask operation earlier in AppInit()

n2yen opened the pull request:
When the option -sysperms=false (default) is enabled, the created datadirs doesn't take into account the default umask setting of '077'.
The fix moves the umask operation to just after parameters are parsed during AppInit()

laanwj commented:
Or, if this is a big issue, maybe do it in both places?

MarcoFalke commented:
Would be nice to know when this regression was introduced. If it never did anything, I guess it can be removed?

If it is not going to be removed, this must come with functional tests that verify the behavior.

n2yen commented:
Appreciate the feedback, I will look into these later - please let me know if you have any other suggestion as well.  Thank you!

n2yen commented:
Had a look at the bitcoin-qt as suggested - it looks like if there was a regression it could be because it potentially generates datadir on its own, without calling GetDefaultDataDir(), see comments in the commit above. Tested by running bitcoind, and bitcoin-qt and checking datadir. The existing unit tests, don't appear to rely on either of the above utility function calls - should I still go ahead and add unit tests for this anyway? thanks!

ken2812221 commented:
@n2yen You can add a python script into functional test, it allow you to run a node to test this.

laanwj commented:
> If it is not going to be removed, this must come with functional tests that verify the behavior.

It definitely works, but not for the initial directory creation. It works for files created after that, and that's what it's intended for (block file data sharing, basically).

n2yen commented:
@Iaanwj can you clarify if we should also leave the old the old code where it was also? i'm not sure, thank you for your feedback!

MarcoFalke commented:
@n2yen I haven't looked, but according to the issue  #13371 the old code is doing nothing. Having useless code is useless, so please remove if it is useless.

laanwj commented:
So to show the current code should not naively be removed (nor applied for `bitcoind` only), this affects wallet default permissions:
```bash
$ src/bitcoind -datadir=/tmp/wallet1 -regtest
$ stat /tmp/wallet1/regtest/wallets/wallet.dat 
Access: (0600/-rw-------)
```
After removing the `umask` setting code from `AppInitBasicSetup`:
```bash
$ src/bitcoind -datadir=/tmp/wallet2 -regtest
$ stat /tmp/wallet2/regtest/wallets/wallet.dat
Access: (0660/-rw-rw----)
```
(wallet is group-readable now)

n2yen commented:
@Iaanwj ok, so we're in agreement, that the umask in AppInitBasicSetup() may not be early enough.  I agree that the umask cannot just be in bitcoind.cpp, but also in bitcoin-qt or some common code that can be called earlier... I will also be adding tests in the way suggested by @ken2812221.

n2yen committed:
Fix #13371 - call umask earlier and remove it from AppInitBasicSetup() as it is too late there.

When the option -sysperms=false (default) is enabled, the created datadirs does not take into account the default umask setting of '077'.
Fix moves the umask operation as early as possible inside of bitcoind, and bitcoin-qt. This was needed because by the time AppInitBasicSetup()
is called, datadir, has typically already been created. Calling umask earlier ensures any filesystem writes that follows has the correct umask set.

Tests: feature_sysperms.py - validates datadir permissions where -sysperms=false (default)
Notes:
'-sysperms' option requires '-disablewallet'
- add check for windows platform, feature_nosysperms.py will explicitly skip the test if
  'Windows' platform is detected.

if bitcoind is built with NO_WALLET=1. That is -sysperms will fail if
not accompanied with -disablewallet option

n2yen commented:
Updated commit with tests to verify permissions of regtest directory, and umask is set as early as sensible in both bitcoind and bitcoin-qt.

n2yen commented:
Rebased on top of:
eb2f1bd276108e70aff0f582a407e9b702eb4dd1 - Merge #14189: qa: Fix silent merge conflict in wallet_importmulti...
On sept 10, 2018

practicalswift commented on test/functional/feature_nosysperms.py:
```diff
@@ -0,0 +1,43 @@
+
+    def check_permissions(self, fullpathname):
+        self.log.info('{}  {}'.format(stat.filemode(os.stat(fullpathname).st_mode), fullpathname))
+        return (os.stat(fullpathname).st_mode & (stat.S_IRWXO|stat.S_IRWXG)) == 0
```
```
2018-09-14 14:23:42 flake8(pr=13389): test/functional/feature_nosysperms.py:19:62: E227 missing whitespace around bitwise or shift operator
```

n2yen replied:
Done.

practicalswift commented on test/functional/feature_nosysperms.py:
```diff
@@ -0,0 +1,43 @@
+
+    def run_test(self):
+        #self.stop_node(0)
+        #self.start_node(0, ['-regtest'])
```
Remove these two comments :-)

n2yen replied:
Done. Thanks for the feedback!

ken2812221 commented:
Concept ACK. 
nit: You should skip the test on Windows.

n2yen commented:
Hi @ken2812221, 
Thanks for your feedback. I've made the changes, to skip the test on Windows, however I don't have a windows machine to confirm it skips. Can someone try it? I've confirmed the tests will run on Linux.

n2yen commented:
Thanks Ken2812221, yup, I'll look into it.

practicalswift commented:
@n2yen Job restarted :-)

n2yen commented:
@practicalswift thank you!

hebasto commented on test/functional/feature_sysperms.py:
```diff
@@ -0,0 +1,60 @@
+        self.restart_node(0)
+        datadir = self.nodes[0].datadir
+
+        self.log.info("Running test without -sysperms option - assuming umask of '0o77'")
```
`'0o77'` - typo?

n2yen replied:
you're right, should be '077'. What did you make of -sysperm option, is it still needed?

hebasto replied:
> What did you make of -sysperm option, is it still needed?

My concerns about `-sysperms` are in the note here #15902.

n2yen replied:
does it makes sense to close this PR - perhaps replace it with simply removing the option? i suppose the tests might still be useful...

MarcoFalke commented:
Please squash your commits according to https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits

KaosGhostNINJA commented:
I believe  that it might not work on windows till I finish what I am doing.

On Mon, May 20, 2019, 12:57 PM MarcoFalke <notifications@github.com> wrote:

> Please squash your commits according to
> https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/13389?email_source=notifications&email_token=AKW4MATWPOMKXUYQ25U6F4TPWLRCBA5CNFSM4FDB353KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVZTA5Y#issuecomment-494088311>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AKW4MAWEAIIQZMY3QT6RPETPWLRCBANCNFSM4FDB353A>
> .
>

n2yen commented:
rebased!

jonasschnelli commented:
utACK 09072f762ae3fe9a746ab447458e95a6fb646ed3

laanwj commented:
I don't think my original comments have been addressed. This still doesn't allow setting sysperm in the configuration file, only on the command line.

It also complicates the initialization order.

I'm starting to think it would be better to remove the `sysperm` setting.

jonasschnelli commented:
I underestimated the fact that one can no longer use the `sysperm` setting in the config file after this. I just retrieved my ACK.

jonasschnelli commented:
@n2yen: what is the status of this PR? Can you have a look in addressing/commenting @laanwj [comment](https://github.com/bitcoin/bitcoin/pull/13389#issuecomment-539906114)?

n2yen commented:
ok, will have a look.

n2yen committed:
rebased and move sysperms check after reading of config file

jonasschnelli commented:
Closing in favour of #17127 (which is probably superior).
